File: Python/Karakter.py
from Factions import Factions

class Karakter:
    """Mencakup semua karakter yang ada di game ini (Base Class)."""
    def __init__(self, name: str, gender: str, race: str, affiliation: Factions = None, **kwargs):
        # Menerima **kwargs tapi mengabaikannya karena ini adalah kelas dasar
        self.name = name
        self.gender = gender
        self.race = race
        self.affiliation = affiliation

    def display(self):
        print(f"Nama: {self.name}")
        print(f"  Gender: {self.gender}")
        print(f"  Ras: {self.race}")
        
        aff_name = self.affiliation.name if self.affiliation else "Tidak Berafiliasi"
        print(f"  Afiliasi: {aff_name}")

File: C++/Karakter.cpp
#pragma once
#include <iostream>
#include <string>
#include "Factions.cpp"

class Karakter {
protected:
    std::string name;
    std::string gender;
    std::string race;
    Factions* affiliation; // Aggregation (0..1)

public:
    // Konstruktor ini akan dipanggil oleh Companion (most derived class)
    Karakter(std::string name, std::string gender, std::string race, Factions* affiliation = nullptr)
        : name(name), gender(gender), race(race), affiliation(affiliation) {}

    virtual ~Karakter() = default;

    // Getter
    std::string getName() const { return name; }
    std::string getGender() const { return gender; }
    std::string getRace() const { return race; }
    Factions* getAffiliation() const { return affiliation; }

    // Setter
    void setName(std::string n) { name = n; }
    void setGender(std::string g) { gender = g; }
    void setRace(std::string r) { race = r; }
    void setAffiliation(Factions* f) { affiliation = f; }

    // Display (virtual)
    virtual void display() const {
        std::cout << "Nama: " << name << std::endl;
        std::cout << "  Gender: " << gender << std::endl;
        std::cout << "  Ras: " << race << std::endl;
        std::cout << "  Afiliasi: ";
        if (affiliation) {
            std::cout << affiliation->getName() << std::endl;
        } else {
            std::cout << "Tidak Berafiliasi" << std::endl;
        }
    }
};